### PR TITLE
Fix solver! when T=Float32

### PR DESCRIPTION
**Pull request by weymouth**
I have found a few dynamic body cases (including the new Optim example) where `solver!` diverges (or nearly diverges) when using `T=Float32` (or `Dual{Float32}`). I _think_ this is because the inner product within PCG is not super accurate when using Float32, especially on a deep MG level with small residuals, so you can end up adding noise instead of making progress. 

A _good_ solution would be to improve the inner product function. That's not trivial, but it would be a good thing to look into. 

What I've done instead is stabilize the solver a little by checking if the preconditioned residual `r'*D^{-1}*r` is increasing instead of decreasing (as it should). This is a good indication that somethings going wrong, and you should stop working at this level. I've set a tolerance of 10% increase for GMG smoothing, and 100% for pure PCG solving. 

I also created an extra parameter `ml.maxlevels` to limit the number of levels in a Vcycle, avoiding recreating levels when incrementing down.

### CODE REVIEW

**Comment by weymouth** [comment]
I would also like to update the optimization example in this PR, but I think the updated solver is ready for review.

File: src/Poisson.jl [hunk 1]
```diff
@@ -163,7 +163,7 @@ function solver!(p::Poisson;log=false,tol=1e-4,itmx=1e3)
     log && (res = [r₂])
     nᵖ=0
     while r₂>tol && nᵖ<itmx
-        smooth!(p); r₂ = L₂(p)
+        pcg!(p); r₂ = L₂(p)
```
Comment (b-fg) [inline src/Poisson.jl:166]: why this change?

**Comment by b-fg** [comment]
I had a look at it and all seems fine to me. But maybe @marinlauber can also doube-check it.
About improving the inner product, what are you thinking of? Isn't just the problem that the noise of `Float32` is at the level of the divergence-free criterion? Wouldn't the casting to `Float64` help here?

**Comment by weymouth** [comment]
We can't cast because we don't do this with a working array, like `df` in `pressure force`. PCG is already using every single scalar field in the code. We could add another for this purpose and see if it helps...

**Comment by b-fg** [comment]
I meant to cast within a custom kernel. I will play with it and see if that works.

**Comment by weymouth** [comment]
Give it a try.

File: src/Poisson.jl (hunk 1 above)
Comment (weymouth) [reply src/Poisson.jl:166]: Because when PCG is used directly as a solver (not as a smoother) it can get stuck without making any progress if we don't allow a larger error margin. The GMG should just quit and try something else.

**Comment by b-fg** [comment]
Made a test to compare a custom dot product using `mapreduce` vs `LinearAlgebra.⋅`. It performs well on GPU, but the allocation size scales with array size `N` on CPU. This is not coming from the type promotion.
```julia
using CUDA
using LinearAlgebra: ⋅

function dot_mapreduce64(a, b)
    mapreduce(+, a, b) do x, y
        promote_type(Float64,eltype(a))(x*y)
    end
end

N = (100,1000,1000)
v = rand(Float32,N)
v_gpu = v |> CuArray
println(v⋅v)
println(v_gpu⋅v_gpu)
println(dot_mapreduce64(v,v))
println(dot_mapreduce64(v_gpu,v_gpu))

@time v⋅v
CUDA.@time v_gpu⋅v_gpu
@time dot_mapreduce64(v,v)
CUDA.@time dot_mapreduce64(v_gpu,v_gpu)
```
```
333715.56
333717.56
333717.5831831272
333717.5831831272

0.000246 seconds (1 allocation: 16 bytes)
0.000065 seconds (15 CPU allocations: 240 bytes)
0.001476 seconds (3 allocations: 7.629 MiB)
0.000368 seconds (163 CPU allocations: 5.016 KiB) (2 GPU allocations: 200 bytes, 10.66% memmgmt time)
```

**Comment by weymouth** [comment]
This can't be a problem unique to us. What is the difference between the LinearAlgebra routine and ours?

**Comment by b-fg** [comment]
For the CPU arrays, LinearAlgebra has a simple method of `dot`:
https://github.com/JuliaLang/julia/blob/4e1fd72042133f0dfdcc3f4c3ce5fa74f76bb9c5/stdlib/LinearAlgebra/src/generic.jl#L957
But for the GPU arrays, they just offer an interface to CUBLAS.

Last resort, we could implement a `@simd` version of this in CPU and extend this in the CUDA extension file with the propose method. Also, I have asked in Slack to see if someone comes up with an explanation for the allocations of `mapreduce` in CPU.

**Comment by marinlauber** [comment]
Can we not use something like the Poisson `mult` routine? They are not allocating on either CPU/GPU, but I don't know about the type conversion here
```
@fastmath @inline function ⋅(a::AbstractArray{T},b::AbstractArray{T}) where T
    s = zero(promote_type(Float64,T))
    for I ∈ inside(a)
        @inbounds s += a[I]*b[I]
    end
    return s
end
```

**Comment by b-fg** [comment]
This does not work on GPU for me, rightfully complains about indexing.

**Comment by b-fg** [comment]
I suggest including the following in `utils.jl`
```julia
function ⋅(a, b)
    @assert size(a) == size(b) "`size(a)` and `size(b)` are not matching."
    s = zero(promote_type(Float64,eltype(a)))
    @simd for i ∈ eachindex(a)
        @inbounds s += a[i]*b[i]
    end
    s
end
```
which is as fast and non-allocating (just a double float) as `LinearAlgebra: ⋅`. Then we can reimplement it in the CUDA extension with
```julia
function ⋅(a, b)
    @assert size(a) == size(b) "`size(a)` and `size(b)` are not matching."
    mapreduce(+, a, b) do x, y
        promote_type(Float64,eltype(a))(x*y)
    end
end
```
This one is 3 times slower than `LinearAlgebra: ⋅` (which uses CUBLAS for CuArrays), but allows us to reduce using `Float64`. Thought @weymouth ?
